File: mcd_dart/lib/generate_hashes.dart
import 'dart:io';
import 'package:args/args.dart';
import 'package:fast_log/fast_log.dart';
import 'package:path/path.dart' as path;
import 'package:mcd_dart/mcd_dart.dart';

/// Utility to generate perceptual hash database from reference card images
void main(List<String> arguments) async {
  // Parse command-line arguments
  ArgParser parser = ArgParser()
    ..addOption('input', abbr: 'i', help: 'Directory containing reference card images', mandatory: true)
    ..addOption('output', abbr: 'o', help: 'Path to output hash file', defaultsTo: Config.getDefaultReferenceHashPath())
    ..addOption('set', abbr: 's', help: 'Set name (e.g., "alpha", "beta", "unlimited")', defaultsTo: 'custom')
    ..addFlag('verbose', abbr: 'v', help: 'Enable verbose output', defaultsTo: false)
    ..addFlag('help', abbr: 'h', help: 'Show this help message', defaultsTo: false);

  try {
    ArgResults results = parser.parse(arguments);

    if (results['help'] as bool) {
      printUsage(parser);
      exit(0);
    }

    // Get parameters
    String inputDir = results['input'] as String;
    String outputFile = results['output'] as String;
    String setName = results['set'] as String;
    bool verbose = results['verbose'] as bool;
    
    // If using default output path, modify it to include the set name
    if (outputFile == Config.getDefaultReferenceHashPath()) {
      final hashesDir = Config.getSetHashesDirectory();
      outputFile = path.join(hashesDir, '${setName}_reference_phash.dat');
      info('Using output file: $outputFile');
    }

    // Check if input directory exists
    final dir = Directory(inputDir);
    if (!await dir.exists()) {
      error('Error: Input directory does not exist: $inputDir');
      exit(1);
    }

    // Initialize the card detector
    MagicCardDetector detector = MagicCardDetector();
    detector.verbose = verbose;

    info('Reading reference images from $inputDir');
    try {
      await detector.readAndAdjustReferenceImages(inputDir);
      info('Loaded ${detector.referenceImages.length} reference images');
    } catch (e) {
      error('Error reading reference images: $e');
      exit(1);
    }

    // Export hash data
    info('Exporting reference hash data to $outputFile');
    try {
      // Create output directory if it doesn't exist
      final outputDir = Directory(path.dirname(outputFile));
      if (!await outputDir.exists()) {
        await outputDir.create(recursive: true);
        info('Created output directory: ${outputDir.path}');
      }
      
      await detector.exportReferenceData(outputFile);
      info('Hash database successfully exported with ${detector.referenceImages.length} cards');
    } catch (e) {
      error('Error exporting hash data: $e');
      exit(1);
    }

    success('Done!');
  } catch (e) {
    error('Error: $e');
    printUsage(parser);
    exit(1);
  }
}

void printUsage(ArgParser parser) {
  info('Magic Card Detector - Hash Generator');
  info('This tool generates a perceptual hash database from reference card images.');
  info('The hash database can be used by the detector for faster card recognition.');
  info('Usage: generate_hashes [options]');
  info(parser.usage);
}

File: MCD_Python/bin/generate_hashes.py
#!/usr/bin/env python
"""
Hash generation tool for MTG card detector.

This tool generates perceptual hashes for a set of MTG card images
and saves them to a reference file.
"""

import os
import argparse
import cv2
from mtg_card_detector import MagicCardDetector


def main():
    """
    Main function for the hash generation tool.
    """
    parser = argparse.ArgumentParser(
        description='Generate perceptual hashes for MTG card images. ' +
                   'Author: Timo Ikonen, timo.ikonen(at)iki.fi')

    parser.add_argument('--set-path', '-p', required=True,
                      help='Path to the folder containing card images.')
    parser.add_argument('--output', '-o', required=True,
                      help='Output file path for the hash data.')
    parser.add_argument('--verbose', '-v', action='store_true',
                      help='Enable verbose output.')

    args = parser.parse_args()

    # Validate the set path
    set_path = args.set_path
    if not os.path.isdir(set_path):
        print(f"Error: Set path '{set_path}' is not a valid directory.")
        return 1
    
    # Add trailing slash if needed
    if not set_path.endswith('/'):
        set_path += '/'

    # Create the output directory if it doesn't exist
    output_dir = os.path.dirname(args.output)
    if output_dir and not os.path.exists(output_dir):
        os.makedirs(output_dir)

    print(f"Generating hashes for card images in {set_path}")
    print(f"Output will be saved to {args.output}")

    # Create detector and generate hashes
    detector = MagicCardDetector()
    detector.verbose = args.verbose
    
    # Load reference images and generate hashes
    try:
        detector.read_and_adjust_reference_images(set_path)
        
        if not detector.reference_images:
            print(f"No .jpg images found in {set_path}")
            return 1
            
        print(f"Found {len(detector.reference_images)} images")
        
        # Export the hash data
        detector.export_reference_data(args.output)
        print(f"Hash data successfully saved to {args.output}")
        
        return 0
    except Exception as e:
        print(f"Error generating hashes: {e}")
        return 1


if __name__ == "__main__":
    main()